File: 20. Dynamic Programming/target_sum.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int s = 0;
        
        for(auto x : nums)
            s += x;
             
        //count number of subsets with sum as their summation
       if(s < target || (target + s)&1 )
           return 0;
            
        int n = nums.size();
        int sum = (abs(target) + s)/2;
        int dp[n+1][sum+1];
        
        for(int i=0;i<=n;i++)
            for(int j=0;j<=sum;j++)
            {
                if(i == 0 && j == 0)
                    dp[i][j] = 1;
                
                else if(i == 0)
                    dp[i][j] = 0;
                
                else if(j < nums[i-1])
                    dp[i][j] = dp[i-1][j];
                
                else
                    dp[i][j] = dp[i-1][j] + dp[i-1][j-nums[i-1]];
            }
        
        return dp[n][sum];
    }
};
/*
this problem is similar of just same as the SUBSET SUM PROBLEM
*/

File: 20. Dynamic Programming/TargetSum.java
public class TargetSum{
  
   public static void main(String args[]){
     int nums[]={1,3,5,2,6};
     int target=4;
     System.out.print(findTargetSumWays(nums,target));
   }
    public static int findTargetSumWays(int[] nums, int target) {
       HashMap<String,Integer> mp=new HashMap<>();
       return target(nums,target,0,mp);
        
    }
   static int target(int arr[],int sum,int ind,HashMap<String,Integer> mp){
        if(ind>=arr.length){
            if(sum==0) return 1;
            return 0;
        }
       String key=ind+":"+sum;
       if(mp.containsKey(key)){
           return mp.get(key);
       } 
       int left=target(arr,sum-arr[ind],ind+1,mp);
       int right=target(arr,sum+arr[ind],ind+1,mp);
       
          mp.put(key,left+right);
            return mp.get(key);
        
    }
  
}
